sdk/python/feast/image_utils.py

```diff
@@ -0,0 +1,279 @@
+        """
+        _check_image_dependencies()
+
+        try:
+            self.model_name = model_name
+            self.model = timm.create_model(
+                model_name, pretrained=True, num_classes=0, global_pool="avg"
+            )
+            self.model.eval()
+
+            config = resolve_data_config({}, model=model_name)
+            self.preprocess = create_transform(**config)
+        except Exception as e:
+            raise RuntimeError(f"Failed to load model '{model_name}': {e}")
+
+    def extract_embedding(self, image_bytes: bytes) -> List[float]:
+        """
+        Extract embedding from image bytes.
+        Args:
+            image_bytes: Image data as bytes (JPEG, PNG, WebP, etc.)
+        Returns:
+            Normalized embedding vector as list of floats
+        Raises:
+            ValueError: If image cannot be processed or is invalid
+        """
+        try:
+            image = Image.open(io.BytesIO(image_bytes)).convert("RGB")
+            input_tensor = self.preprocess(image).unsqueeze(0)
+
+            with torch.no_grad():
+                output = self.model(input_tensor)
+
+            feature_vector = output.squeeze().numpy()
+            normalized = normalize(feature_vector.reshape(1, -1), norm="l2")
+            return normalized.flatten().tolist()
+
+        except Exception as e:
+            raise ValueError(f"Failed to extract embedding from image: {e}")
+
+    def get_embedding_dimension(self) -> int:
+        """
+        Get the dimension of embeddings produced by this model.
+        Returns:
+            Integer dimension of the embedding vector
+        """
+        dummy_input = torch.randn(1, 3, 224, 224)  # Standard input size
+        with torch.no_grad():
+            output = self.model(dummy_input)
+        return output.shape[1]
+
+    def batch_extract_embeddings(
+        self, image_bytes_list: List[bytes]
+    ) -> List[List[float]]:
+        """
+        Extract embeddings from multiple images in batch for efficiency.
+        Args:
+            image_bytes_list: List of image data as bytes
+        Returns:
+            List of normalized embedding vectors
+        Raises:
+            ValueError: If any image cannot be processed
+        """
+        embeddings = []
+
+        images = []
+        for image_bytes in image_bytes_list:
+            try:
+                image = Image.open(io.BytesIO(image_bytes)).convert("RGB")
+                preprocessed = self.preprocess(image)
+                images.append(preprocessed)
+            except Exception as e:
+                raise ValueError(f"Failed to preprocess image: {e}")
+
+        batch_tensor = torch.stack(images)
+
+        with torch.no_grad():
+            outputs = self.model(batch_tensor)
+
+        for output in outputs:
+            feature_vector = output.numpy()
+            normalized = normalize(feature_vector.reshape(1, -1), norm="l2")
```

Comment:
feel like l2 should be configurable but probably not needed yet

Reply:
right, most of the usecase get covered via l2, we can expand as needed